src/gc/gc.d

```diff
@@ -1671,6 +1680,75 @@ struct Gcx
         return pool.npages * PAGESIZE;
     }
 
+    /**
+     * Update the thresholds for when to collect the next time
+     */
+    void updateCollectThresholds() nothrow
+    {
+        static float max(float a, float b) nothrow
+        {
+            return a >= b ? a : b;
+        }
+
+        static size_t roundUp(float val, size_t multiple) nothrow
+        {
+            return cast(size_t)(val + multiple - 1 + 0.5) / multiple * multiple;
+        }
+
+        static float zeroDenormal(float val) nothrow
+        {
+            return val <= float.min_normal ? 0 : val;
+        }
+
+        static float decay(float oldVal, float newVal) nothrow
+        {
+            // one-pole RC low-pass
+            // see http://en.wikipedia.org/wiki/RC_time_constant
+            // and http://en.wikipedia.org/wiki/Low-pass_filter#Simple_infinite_impulse_response_filter
+
+            // decay to 63.2% of newVal over 5 collections
+            enum alpha = 1.0 / (5 + 1);
+            return zeroDenormal((newVal - oldVal) * alpha + oldVal);
+        }
+
+        // steep attack, smooth decay
+        immutable smTarget = usedSmallPages * GC.config.heapSizeFactor;
+        smallCollectThreshold = max(smTarget, decay(smallCollectThreshold, smTarget));
+        immutable lgTarget = usedLargePages * GC.config.heapSizeFactor;
+        largeCollectThreshold = max(lgTarget, decay(largeCollectThreshold, lgTarget));
+    }
+
+    void initMemLimit() nothrow
+    {
+        enum GB = 2UL ^^ 30;
+        enum perCent = 80;
+
+        version (D_LP64) {}
+        else version (Windows)
+        {
+            import core.sys.windows.windows;
+            MEMORYSTATUS stat;
+            GlobalMemoryStatus(&stat);
+            vmemLimit = stat.dwTotalPhys * perCent / 100;
```

Comment:
Physical memory is not the proper metric, these days virtual address space is more scarce. We might use the minimum of both. The "available virtual memory" might also be useful.

I'd expect a function to retreive these limits in gc/os.d.